browser/ntp_background/ntp_p3a_helper_impl_unittest.cc

```diff
@@ -0,0 +1,156 @@
+  ASSERT_TRUE(p3a_service_->IsDynamicMetricRegistered(histogram_name));
+
+  ntp_p3a_helper_->OnP3AMetricSent(histogram_name);
+
+  histogram_tester_->ExpectTotalCount(histogram_name, 2);
+  ASSERT_FALSE(p3a_service_->IsDynamicMetricRegistered(histogram_name));
+}
+
+TEST_F(NTPP3AHelperImplTest, OneEventTypeCountReportedWhileInflight) {
+  ntp_p3a_helper_->RecordClickAndMaybeLand(kTestCreativeMetricId);
+  ntp_p3a_helper_->RecordClickAndMaybeLand(kTestCreativeMetricId);
+
+  const std::string histogram_name = GetExpectedHistogramName("clicks");
+
+  ASSERT_TRUE(p3a_service_->IsDynamicMetricRegistered(histogram_name));
+
+  histogram_tester_->ExpectTotalCount(histogram_name, 0);
+
+  ntp_p3a_helper_->OnP3ARotation(true);
+
+  histogram_tester_->ExpectBucketCount(histogram_name, 2, 1);
+
+  // Recorded a click while recorded count is "in-flight"
+  ntp_p3a_helper_->RecordClickAndMaybeLand(kTestCreativeMetricId);
+  ntp_p3a_helper_->OnP3AMetricSent(histogram_name);
+
+  ASSERT_TRUE(p3a_service_->IsDynamicMetricRegistered(histogram_name));
+
+  ntp_p3a_helper_->OnP3ARotation(true);
+  histogram_tester_->ExpectBucketCount(histogram_name, 1, 1);
+
+  ntp_p3a_helper_->OnP3AMetricSent(histogram_name);
+
+  ASSERT_FALSE(p3a_service_->IsDynamicMetricRegistered(histogram_name));
+  histogram_tester_->ExpectTotalCount(histogram_name, 2);
+}
+
+TEST_F(NTPP3AHelperImplTest, LandCountReported) {
+  ntp_p3a_helper_->RecordClickAndMaybeLand(kTestCreativeMetricId);
+
+  const std::string histogram_name = GetExpectedHistogramName("lands");
+
+  ASSERT_FALSE(p3a_service_->IsDynamicMetricRegistered(histogram_name));
+
+  ntp_p3a_helper_->SetLastTabURL(GURL("https://adexample.com/page1"));
+
+  task_environment_.FastForwardBy(base::Seconds(6));
+
+  // It's acceptable to access other pages,
+  // as long as they're on the same host.
+  ntp_p3a_helper_->SetLastTabURL(GURL("https://adexample.com/page2"));
+
+  task_environment_.FastForwardBy(base::Seconds(5));
+
+  histogram_tester_->ExpectTotalCount(histogram_name, 0);
+
+  ntp_p3a_helper_->OnP3ARotation(true);
+  ASSERT_TRUE(p3a_service_->IsDynamicMetricRegistered(histogram_name));
+  histogram_tester_->ExpectBucketCount(histogram_name, 1, 1);
+
+  ntp_p3a_helper_->RecordClickAndMaybeLand(kTestCreativeMetricId);
+
+  ntp_p3a_helper_->SetLastTabURL(GURL("https://adexample.com/page1"));
+
+  task_environment_.FastForwardBy(base::Seconds(6));
+  histogram_tester_->ExpectTotalCount(histogram_name, 1);
+
+  // Should not trigger land, since user left page before "land time"
+  ntp_p3a_helper_->SetLastTabURL(GURL("https://differenthost.com/page1"));
+
+  task_environment_.FastForwardBy(base::Seconds(5));
+
+  ntp_p3a_helper_->OnP3ARotation(true);
+  histogram_tester_->ExpectBucketCount(histogram_name, 1, 2);
+
+  ASSERT_TRUE(p3a_service_->IsDynamicMetricRegistered(histogram_name));
+  ntp_p3a_helper_->OnP3AMetricSent(histogram_name);
+  ASSERT_FALSE(p3a_service_->IsDynamicMetricRegistered(histogram_name));
```

Comment:
tests should use EXPECT for expectations

Reply:
any reason why we can't immediately abort, which `ASSERT_` seems to provide?

Reply:
If we change `ASSERT_TRUE` to `EXPECT_TRUE`, the test can show that `IsDynamicMetricRegistered` is implemented incorrectly, but other logic works fine. EXPECT_TRUE is preferred by default.

> Usually EXPECT_* are preferred, as they allow more than one failures to be reported in a test. However, you should use ASSERT_* if it doesn't make sense to continue when the assertion in question fails.

Reply:
> However, you should use ASSERT_* if it doesn't make sense to continue when the assertion in question fails.

This is the desired behavior